File: avalon-modules/avalon-auth/avalon-auth-web/src/main/java/io/github/lishangbu/avalon/auth/controller/TokenController.java
package io.github.lishangbu.avalon.auth.controller;

import io.github.lishangbu.avalon.auth.model.SignInPayload;
import io.github.lishangbu.avalon.auth.model.TokenInfo;
import io.github.lishangbu.avalon.security.util.JwtUtils;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * 令牌控制器
 *
 * @author lishangbu
 * @since 2025/4/9
 */
@RestController
@RequestMapping("/token")
@RequiredArgsConstructor
public class TokenController {
  private final AuthenticationManager authenticationManager;
  private final JwtUtils jwtUtils;

  @PostMapping("/sign-in")
  public TokenInfo signIn(@RequestBody @Valid SignInPayload signInPayload) {
    Authentication authentication =
        authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(
                signInPayload.getUsername(), signInPayload.getPassword()));
    SecurityContextHolder.getContext().setAuthentication(authentication);
    return new TokenInfo(
        jwtUtils.generateAccessTokenByAuthentication(authentication),
        jwtUtils.generateRefreshTokenByAuthentication());
  }
}

File: avalon-modules/avalon-authorization/src/main/kotlin/io/github/lishangbu/avalon/authorization/controller/TokenController.kt
package io.github.lishangbu.avalon.authorization.controller

import io.github.lishangbu.avalon.oauth2.common.userdetails.UserInfo
import org.springframework.security.core.Authentication
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.core.OAuth2AccessToken
import org.springframework.security.oauth2.server.authorization.OAuth2Authorization
import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService
import org.springframework.security.oauth2.server.authorization.OAuth2TokenType
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

/**
 * 令牌控制器
 *
 * 提供登出与当前用户令牌信息查询接口
 *
 * @author vains
 */
@RequestMapping("/token")
@RestController
class TokenController(
    /** OAuth2 授权服务 */
    private val oAuth2AuthorizationService: OAuth2AuthorizationService,
) {
    /** 注销当前登录会话 */
    @DeleteMapping("/logout")
    fun logout() {
        val authentication: Authentication? = SecurityContextHolder.getContext().authentication
        if (authentication != null && authentication.credentials is OAuth2AccessToken) {
            val accessToken = authentication.credentials as OAuth2AccessToken
            val auth2Authorization: OAuth2Authorization? =
                oAuth2AuthorizationService.findByToken(
                    accessToken.tokenValue,
                    OAuth2TokenType.ACCESS_TOKEN,
                )
            if (auth2Authorization != null) {
                oAuth2AuthorizationService.remove(auth2Authorization)
                SecurityContextHolder.clearContext()
            }
        }
    }

    /** 获取当前登录用户信息 */
    @GetMapping("/info")
    fun user(
        @AuthenticationPrincipal user: UserInfo,
    ): UserInfo = user
}
